sycl/doc/extensions/experimental/sycl_ext_oneapi_spec_constant_length_alloca.asciidoc

```diff
@@ -0,0 +1,266 @@
+specialization constants.
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_SPEC_CONSTANT_LENGTH_ALLOCA` to one of the values defined
+in the table below. Applications can test for the existence of this macro to
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's features the
+implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== New Aspect for Specification Constant-Length Allocations
+
+This extension adds a new device aspect:
+
+[source,c++]
+----
+namespace sycl {
+enum class aspect : /*unspecified*/ {
+  ext_oneapi_scla
+};
+} // namespace sycl
+----
+
+The `ext_oneapi_scla` aspect indicates that the device is capable of using the
+`private_alloca` API defined in the following sections.
+
+=== The SCLA API
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+template <typename ElementType, auto &SpecName,
+          access::decorated DecorateAddress>
+private_ptr<ElementType, DecorateAddress>
+private_alloca(kernel_handler &kh);
+
+template <typename ElementType, std::size_t Alignment, auto &SpecName,
+          access::decorated DecorateAddress>
+private_ptr<ElementType, DecorateAddress>
+aligned_private_alloca(kernel_handler &kh);
+} // namespace sycl::ext::oneapi::experimental
+----
+
+_Mandates_: `ElementType` must be a cv-unqualified trivial type and `SpecName`
+must be a reference to a specialization constant of integral `value_type`. In
+the case of `aligned_private_alloca`, `Alignment` must be a power of 2
+fundamental alignment stricter than the alignment requirement of `ElementType`.
+
+_Effects_: `h.get_specialization_constant<size>()` elements of type
+`ElementType` are allocated and default initialized in private memory.
+
+_Returns_: A pointer to a default initialized region of private memory of
+`h.get_specialization_constant<size>()` elements of type
+`ElementType`. `DecorateAddress` defines whether the returned `multi_ptr` is
+decorated. In the case of `private_alloca`, the pointer is suitably aligned for
+an object of type `ElementType`. In the case of `aligned_private_alloca`, the
+pointer is aligned to the specified `Alignment`.
+
+_Remarks_: In case of private memory exhaustion, the underlying backend must
+report an error in the same fashion as if the allocation size were static. In
+case of a successful call, allocated memory has automatic storage
+duration. Additionally, `SpecName` must have a default value of at least 1 and
+not be set to a value less than 1 during program execution. Violation of these
+conditions results in undefined behaviour.
```

Comment:
This last sentence about `SpecName` seems like a precondition.  Could we move it to a _Precondition_ section instead?  There is no need to say that violation of the precondition results in UB because that is the definition of a precondition.

Reply:
Makes sense.